toolkit/tools/precacher/precacher.go

```diff
@@ -0,0 +1,409 @@
+}
+
+var (
+	app = kingpin.New("precacher", "Pre-hydrate RPM cache for a given set of repo URLs and a RPM snapshot file.")
+
+	logFile       = exe.LogFileFlag(app)
+	logLevel      = exe.LogLevelFlag(app)
+	profFlags     = exe.SetupProfileFlags(app)
+	timestampFile = app.Flag("timestamp-file", "File that stores timestamps for this program.").String()
+
+	outDir            = exe.OutputDirFlag(app, "Directory to download packages into.")
+	snapshot          = app.Flag("snapshot", "Path to the rpm snapshot .json file.").ExistingFile()
+	outputSummaryFile = app.Flag("output-summary-file", "Path to save the summary of packages downloaded").String()
+	repoUrls          = app.Flag("repo-url", "URLs of the repos to download from.").Strings()
+	workerTar         = app.Flag("worker-tar", "Full path to worker_chroot.tar.gz").Required().ExistingFile()
+	buildDir          = app.Flag("worker-dir", "Directory to store chroot while running repo query.").String()
+
+	concurrentNetOps = app.Flag("concurrent-net-ops", "Number of concurrent network operations to perform.").Default(defaultNetOpsCount).Uint()
+)
+
+func main() {
+	app.Version(exe.ToolkitVersion)
+	kingpin.MustParse(app.Parse(os.Args[1:]))
+	logger.InitBestEffort(*logFile, *logLevel)
+
+	prof, err := profile.StartProfiling(profFlags)
+	if err != nil {
+		logger.Log.Warnf("Could not start profiling: %s", err)
+	}
+	defer prof.StopProfiler()
+
+	timestamp.BeginTiming("precacher", *timestampFile)
+	defer timestamp.CompleteTiming()
+
+	// Get the state of the current repo from a snapshot
+	rpmSnapshot, err := rpmSnapshotFromFile(*snapshot)
+	if err != nil {
+		logger.PanicOnError(err)
+	}
+
+	// See what is available online from the upstream repos
+	availablePackages, err := getAllRepoData(*repoUrls, *workerTar, *buildDir)
+	if err != nil {
+		logger.PanicOnError(err)
+	}
+
+	logger.Log.Infof("Pre-caching: Found %d available packages", len(availablePackages))
+	if logger.Log.IsLevelEnabled(logrus.DebugLevel) {
+		for _, pkg := range availablePackages {
+			logger.Log.Debugf("Found package: %s", pkg)
+		}
+	}
+
+	// For each package in the snapshot, check if it is available online and try to download it
+	downloadedPackages, err := downloadMissingPackages(rpmSnapshot, availablePackages, *outDir, *concurrentNetOps)
+	if err != nil {
+		logger.PanicOnError(err)
+	}
+
+	logger.Log.Infof("Pre-caching: Downloaded %d packages into the cache", len(downloadedPackages))
+	err = writeSummaryFile(*outputSummaryFile, downloadedPackages)
+	if err != nil {
+		logger.PanicOnError(err)
+	}
+}
+
+func rpmSnapshotFromFile(snapshotFile string) (rpmSnapshot *repocloner.RepoContents, err error) {
+	err = jsonutils.ReadJSONFile(snapshotFile, &rpmSnapshot)
+	return
+}
+
+// getAllRepoData returns a map of package names to URLs for all packages available in the given repos. It uses
+// a chroot to run repoquery.
+func getAllRepoData(repoUrls []string, workerTar, buildDir string) (namesToUrls map[string]string, err error) {
```

Comment:
Nit-pick: let's try using proper abbreviations:
```suggestion
func getAllRepoData(repoURLs []string, workerTar, buildDir string) (namesToURLs map[string]string, err error) {
```